kernel/src/log_segment.rs

```diff
@@ -4,23 +4,166 @@
 use crate::actions::{get_log_schema, Metadata, Protocol, METADATA_NAME, PROTOCOL_NAME};
 use crate::path::ParsedLogPath;
 use crate::schema::SchemaRef;
-use crate::{DeltaResult, Engine, EngineData, Error, Expression, ExpressionRef, FileMeta};
-
+use crate::snapshot::CheckpointMetadata;
+use crate::utils::require;
+use crate::{
+    DeltaResult, Engine, EngineData, Error, Expression, ExpressionRef, FileSystemClient, Version,
+};
 use itertools::Itertools;
+use std::cmp::Ordering;
+use std::convert::identity;
 use std::sync::{Arc, LazyLock};
+use tracing::warn;
 use url::Url;
 
+/// A [`LogSegment`] represents a contiguous section of the log and is made up of checkpoint files
+/// and commit files. It is built with either [`LogSegment::for_table_changes`], or
+/// [`LogSegment::for_snapshot`], and guarantees the following:
+///     1. Commit file versions will not have any gaps between them.
+///     2. If checkpoint(s) is/are present in the range, only commits with versions greater than the most
+///        recent checkpoint version are retained. There will not be a gap between the checkpoint
+///        version and the first commit version.
+///
+/// [`LogSegment`] is used in both  [`Snapshot`] and in `TableChanges` to hold commit files and
+/// checkpoint files.
+///
+/// [`Snapshot`]: crate::snapshot::Snapshot
 #[derive(Debug)]
 #[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
 pub(crate) struct LogSegment {
+    pub end_version: Version,
     pub log_root: Url,
-    /// Reverse order sorted commit files in the log segment
+    /// Sorted commit files in the log segment
     pub commit_files: Vec<ParsedLogPath>,
-    /// checkpoint files in the log segment.
-    pub checkpoint_files: Vec<ParsedLogPath>,
+    /// Checkpoint files in the log segment.
+    pub checkpoint_parts: Vec<ParsedLogPath>,
 }
 
 impl LogSegment {
+    fn try_new(
+        sorted_commit_files: Vec<ParsedLogPath>,
+        checkpoint_parts: Vec<ParsedLogPath>,
+        log_root: Url,
+        expected_end_version: Option<Version>,
```

Comment:
@zachschuermann  We were checking the expected end version in both cases, so I brought that into `try_new`